build and test documentation improvements (#79)
apache/cassandra-website: site-content/source/modules/ROOT/pages/development/ide.adoc

== Building and IDE Integration

=== Prerequisites

Building Cassandra from source is the first important step in contributing to the Apache Cassandra project.
You'll need to install:

- Java 8, also optionally Java 11 - the most convenient and versatile way of installing Java is using https://sdkman.io/[The Software Development Kit Manager]
- it lets you install multiple Java versions and easily switch between them

- GIT - you can use your package manager to install it

- Ant - you can use your package manager to install it; note that you also need to install `ant-optional` module so for example on Debian-like operating systems you should install do that with the following command:

> [source,shell script]

Reviewer: I'm not convinced the expanded "Prerequisites" section here is adding much value or clarity.

Previously this was (basically) the `You’ll need to install Java 8, Git, and Ant first.` sentence.
My suggestion is just to change that sentence to
```
You will need to install Java (JDK 8 or 11), Git, and Ant (including ant-optional).
```

Reviewer 2: concur

----
sudo apt install -y ant ant-optional
----

=== Getting the source code

The source code for Cassandra is shared on the central Apache Git
repository and organized by branch, one branch for each major version.
You can access the code for the current development branch at

> https://github.com/apache/cassandra

Reviewer: Links should favour gitbox.apache.org
(It's ok to also mention that github is an alternative.)

Reviewer 2: Hmm - til about gitbox.

However, for developing a patch or a new feature, you should fork Cassandra
project first and then clone it from your own fork:

[source, shell script]
----
git clone https://github.com/<your_git_name>/cassandra.git cassandra
----

Then, you should create a branch called after the Jira ticket you want
to work on and reset it to the development branch you plan to apply your
change for:

[source, shell script]
----
cd cassandra
git checkout -b CASSANDRA-12345-4.0

> git reset --hard origin/cassandra-4.0

Reviewer: This page is about getting your IDE working with the source code. It is not about writing patches, so this section is out of scope and should be removed.

Author: ok, I will just add the hint to `patches.adoc`

----

[TIP]
.Hint
====
When working on multiple patches at the same time, instead of cloning
the repository separately for each of them you can set up multiple
working trees for different Cassandra versions from the same repository
using https://git-scm.com/docs/git-worktree[Git worktrees] feature.
====

=== Building Cassandra

Cassandra is a Java project which is built using Ant.
The build file, `build.xml` located in the root of the project content, has various tasks defined (you can list all of them with some short descriptions by running `ant -p`).

Assuming you already have JDK installed somewhere in the system. You should export `JAVA8_HOME` and `JAVA11_HOME` environment variables pointing to corresponding JDK locations. Also, the build certainly uses the Java compiler, which is currently set up in your shell. To make it explicit, set the environment variables as follows:

[source,shell script]
----
export JAVA8_HOME=/path/to/jdk8

> export JAVA11_HOME=/path/to/jdk11

Reviewer: This is not a requirement. I never need to set these (anymore).

Author: yes, indeed, I'll remove that

----

then select the JDK you want to use, for example:

[source,shell script]
----
export JAVA_HOME=$JAVA8_HOME
export PATH=$JAVA_HOME/bin:$PATH
----

If you want to build with Java 11, you need to either add a build property `-Duse.jdk11=true` whenever you run Ant build or just export environment variable:

[source,shell script]
----

> export CASSANDRA_USE_JDK11=true

Reviewer: make this an inline code, like `-Duse.jdk11=true` is, instead of a full code block.

----

otherwise, when you run the build using Java 11, but none of those settings are used, the build script complains.

Now you can build Cassandra with just the default task - just execute `ant` or `ant jar`. This may take a significant amount of time depending on artifacts that have to
be downloaded or the number of classes that need to be compiled. When finished, you can find a JAR file in the build directory, and you can run the DB service by executing the `bin/cassandra` script.

=== Building a distribution

Some tasks you might be interested in are:

- `javadoc` - creates API documentation

> - `javadoc-jar`, `sources-jar` - produce accompanying JARs for the primary Cassandra artifact (also in the build directory)

Reviewer: Javadoc* tasks are not interesting. Who uses them anymore?

- `artifacts` - generate Cassandra distribution in `build/dist` directory and package it to `tar.gz` (with and without sources)

While there is JavaDoc JAR or source JAR in the build directory, Cassandra may fail to start. Those jars are needed to publish Cassandra artifacts in a Maven repository but should be removed if you want to run Cassandra.

> You are allowed to set the final version of artifacts when building by passing `-Dversion=<target-version>`. It can be helpful if you, for example, build your custom fork and want to publish artifacts in your repository. You should take two important things into account - the version string you specify must be parsable by Cassandra (see `CassandraVersion` class). Moreover, the version must be parsable by all Cassandra versions you want to support upgrades from.

Reviewer: outside of scope.

Author: ok, I'll remove

== Building and IDE Integration

=== Prerequisites

Building Cassandra from source is the first important step in contributing to the Apache Cassandra project.

> You will need to install Java (JDK 8 or 11), Git, and Ant (including ant-optional).

Reviewer 2: If you mention installing either JDK 8 or 11, you should indicate how a user decides which to install.

=== Getting the source code

The source code for Cassandra is shared on the central Apache Git
repository and organized by branch, one branch for each major version.
You can access the code for the current development branch at
https://gitbox.apache.org/repos/asf/cassandra.git (the primary location)
or at https://github.com/apache/cassandra (a mirror location).

> However, for developing a patch or a new feature, you should fork Cassandra

Reviewer 2: ```suggestion
However, to develop a patch or a new feature, you should fork the Cassandra
```

project first and then clone it from your own fork:

[source, shell script]
----
git clone https://github.com/<your_git_name>/cassandra.git cassandra
----

=== Building Cassandra

Cassandra is a Java project which is built using Ant.
The build file, `build.xml` located in the root of the project content,
has various tasks defined (you can list all of them with some short
descriptions by running `ant -p`).

The build uses the Java compiler which is currently set up in your

> shell. By default, it is assumed you use Java 8. If you want to build

Reviewer 2: ```suggestion
shell. By default, the build uses Java 8. If you want to build
```

with Java 11, you need to either add a build property `-Duse.jdk11=true`

> whenever you run Ant build or just export environment variable

Reviewer 2: ```suggestion
to your Ant build run, or export the environment variable
```

`CASSANDRA_USE_JDK11=true`. Otherwise, when you run the build using

> Java 11 but none of those settings are used, the build script complains.

Reviewer 2: ```suggestion
`CASSANDRA_USE_JDK11=true`. Otherwise, if you run the build using
Java 11, the build script complains.
```

Now you can build Cassandra with just the default task - just execute
`ant` or `ant jar`. This may take a significant amount of time depending
on artifacts that have to be downloaded or the number of classes that

> need to be compiled. When finished, you can find a JAR file in the build

Reviewer 2: ```suggestion
need to be compiled. When the build completes, you can find a JAR file in the build
```

directory, and you can run the DB service by executing

> the `bin/cassandra` script.

Reviewer 2: ```suggestion
directory, and the database service is started with
the `bin/cassandra` script.
```

=== Building a distribution

Some tasks you might be interested in are:

> - `sources-jar` - produce accompanying JAR for the primary Cassandra

Reviewer 2: ```suggestion
* `sources-jar` - produce accompanying JAR for the primary Cassandra
```

artifact (also in the build directory)

> - `artifacts` - generate Cassandra distribution in `build/dist`

Reviewer 2: ```suggestion
* `artifacts` - generate Cassandra distribution in `build/dist`
```

directory and package it to `tar.gz` (with and without sources)

[TIP]
.Hint
====
Remember to have only one Cassandra JAR artifact in the build directory,

> otherwise the tests or Cassandra itself may fail to start! Multiple

Reviewer 2: ```suggestion
Remember to have only one Cassandra JAR artifact in the build directory;
otherwise, the tests or Cassandra itself may fail to start! Multiple
```

> artifacts there may follow building sources or JavaDoc JARs or building

Reviewer 2: The sentence starting with "Multiple" is unclear. I believe you mean:

If you build more than once, leftover artifacts, such as JavaDoc JARs or source JARs, may still be in the build directory. 
Clean up the build directory to avoid conflicts.

Author: Remember to have only one Cassandra JAR artifact in the build directory;
otherwise, the tests or Cassandra itself may fail to start! If you build
sources or JavaDoc, leftover artifacts may be in the build directory.
The same may happen if you switch between different branches, and build
the project many times.

Clean the project after creating auxiliary artifacts or switching to a
different branch to avoid conflicts!

multiple versions of Cassandra without properly cleaning up the build
directory.
====

> There are other tasks more related to testing, and they are covered in link:testing.adoc[Testing] section.

Reviewer 2: ```suggestion
There are other tasks related to testing, and they are covered in link:testing.adoc[Testing] section.
```

----

=== Building Cassandra

Cassandra is a Java project which is built using Ant.
The build file, `build.xml` located in the root of the project content,
has various tasks defined (you can list all of them with some short
descriptions by running `ant -p`).

The build uses the Java compiler which is currently set up in your
shell. By default, the build uses Java 8. If you want to build
with Java 11, you need to either add a build property `-Duse.jdk11=true`
to your Ant build command or export the environment variable
`CASSANDRA_USE_JDK11=true`. Otherwise, if you run the build using
Java 11, the build script complains.

Now you can build Cassandra with the default task - just execute
`ant` or `ant jar`. This may take a significant amount of time depending
on artifacts that have to be downloaded or the number of classes that
need to be compiled. When the build completes, you can find a JAR file
in the build directory, and the database service can be started with
the `bin/cassandra` script.

=== Building a distribution

Some tasks you might be interested in are:

* `sources-jar` - produce accompanying JAR for the primary Cassandra

> artifact (also in the build directory)

Reviewer: Not sure who uses `sources-jar`. Mostly an artefact for releases.

I would replace this with the `build` and `jar`. (It's ok to mention again `jar`, and state here that it is the default.)

Author: I put there `source-jar` and `javadoc-jar` for completeness, but I'm fine with removing them. I'll leave `build`, `build-test`, `artifacts`, `mvn-install` and `realclean` as they seem to be more useful

* `artifacts` - generate Cassandra distribution in `build/dist`
directory and package it to `tar.gz` (with and without sources)

[TIP]
.Hint
====

> Remember to have only one Cassandra JAR artifact in the build directory;

Reviewer: This is still a bit clumsy, because there are always multiple Cassandra JAR artefacts in the build directory, after any build.

e.g. something like… 
```
Remember to have only one versions of the different built jars in the build directory.
```

Author: No, when you just do build with default `jar` task, you have only one `apache-cassandra*.jar` in the `build` directory - so other versions, but also no sources or javadoc jars. This is extremely important because of this part of `casandra.in.sh`:

```shell
if [ -d $CASSANDRA_HOME/build ] ; then
    #cassandra_bin="$CASSANDRA_HOME/build/classes/main"
    cassandra_bin=`ls -1 $CASSANDRA_HOME/build/apache-cassandra*.jar`

    CLASSPATH="$CLASSPATH:$cassandra_bin"
fi
```

In particular, when `ls -1 $CASSANDRA_HOME/build/apache-cassandra*.jar` returns multiple results, they will be split by new line character and completely ruin the classpath. Perhaps that needs some care in Cassandra code, but such implementation is in existing Cassandra versions thus for the documentation purposes it is safe to say that we can only have a single `apache-cassandra*.jar` in the `build` directory. When there are more, you get `Error: Could not find or load main class org.apache.cassandra.service.CassandraDaemon` and you would  be probably like what the heck is wrong, because everything seems to be ok - project is properly built, tests are passing, all jars were generated...

Reviewer: `ant artifacts` generates the following `apache-cassandra*.jar` files…
```
❯ ls build/apache-cassandra*.jar
build/apache-cassandra-4.0.2-SNAPSHOT-javadoc.jar 
build/apache-cassandra-4.0.2-SNAPSHOT-sources.jar 
build/apache-cassandra-4.0.2-SNAPSHOT.jar
```

> 

Reviewer 2: Use menu:File[Open] for `File > Open`. It's a nifty UI macro in antora.

Author: Well, I don't use Antora. I wrote that AsciiDoc in IntelliJ, it has a nice plugin for that

Reviewer 2: antora is the static site generator that is used to build the cassandra-website. It's probably an AsciiDoc feature that antora implements.

* Study Cassandra code style
* Inspections

[TIP]
.Hint
====
If you wish to work with older Cassandra versions, see our https://cwiki.apache.org/confluence/display/CASSANDRA2/RunningCassandraInIDEA[wiki] for instructions.
====

=== Opening Cassandra in Apache NetBeans

> https://netbeans.apache.org/[Apache NetBeans] is an older open source Java IDE,

Reviewer: ```suggestion
https://netbeans.apache.org/[Apache NetBeans] is the elder of the open sourced java IDEs,
```

would you mind returning this to how it was written. I know it isn't part of your changes, but it was recently changed with the new website and lost the nuance that I wrote it with ;) 
see here for the original version: https://cassandra.apache.org/doc/4.0-alpha1/development/ide.html#opening-cassandra-in-apache-netbeans

. Confirm `Debug` and you should see the output of Cassandra start up in the Eclipse console.

You can now set breakpoints and start debugging!

=== General notes

> You may sometimes encounter some weirdly-looking build failures when running those tasks. If you do, please start with running

Reviewer 2: ```suggestion
You may sometimes encounter some odd build failures when running the `ant` commands above. If you do, start `ant` with the `realclean` option:
```

[source, shell script]
----
ant realclean
----

> to clean the project - that usually solves false problems.

Reviewer 2: delete this line

> Remember that all the tasks mentioned above may depend on building source files (this may have changed by the time you read it). If there are actual compilation errors in the code, you may not generate project files for IntelliJ Idea, Netbeans or Eclipse. Especially make sure you have the project imported adequately into IDE before doing merges / rebases because otherwise if there are conflicts, the project cannot be opened in IDE. Thus, you are unable to use any fancy conflict resolution tools offered by those IDEs.

Reviewer 2: what does "(this may have changed by the time you read it)" mean? I would delete this part, unless there is an important statement that you can make about the comment.

Reviewer 2: ```suggestion
Remember that all the tasks mentioned above may depend on building source files. If there are actual compilation errors in the code, you may not be able to generate project files for IntelliJ Idea, Netbeans, or Eclipse.
It is especially important that you have imported the project adequately into IDE before doing merges or rebases. Otherwise, if there are conflicts, the project cannot be opened in IDE, and you will be unable to use any fancy conflict resolution tools offered by those IDEs.
```
